File: cluster-health/src/main/java/org/example/business/MetricsService.java
package org.example.business;

import java.util.List;

import org.example.dataaccess.ClusterOverview;
import org.example.dataaccess.DatabaseMetrics;
import org.example.dataaccess.MetricCard;
import org.example.dataaccess.NamespaceMetric;
import org.example.dataaccess.NodeMetric;
import org.example.dataaccess.ResourceUtilization;
import org.example.dataaccess.UnhealthyPod;


public interface MetricsService {

    // Overview metrics
    ClusterOverview getClusterOverview();

    // Node metrics
    List<NodeMetric> getNodeMetrics();

    // Namespace metrics
    List<NamespaceMetric> getNamespaceMetrics();

    // Unhealthy pods
    List<UnhealthyPod> getUnhealthyPods();

    // Time-series data for charts
    ResourceUtilization getCpuUtilization(String duration);

    ResourceUtilization getMemoryUtilization(String duration);

    // Metric cards with sparklines
    MetricCard getPodCountMetric(String duration);

    MetricCard getNodeCountMetric(String duration);

    MetricCard getRequestCountMetric(String duration);

    // Database metrics
    DatabaseMetrics getDatabaseIoMetrics(String duration);

    MetricCard getTenantCountMetric(String duration);
}

File: observability-service/internal/core/ports/metrics_service.go
package ports

import (
	"context"

	"prometheus-metrics-api/internal/core/domain"
)

type MetricsService interface {
	GetClusterOverview(ctx context.Context) (*domain.ClusterOverview, error)
	GetRequestsMetric(ctx context.Context, duration string) (*domain.MetricCard, error)
	GetPodsMetric(ctx context.Context, duration string) (*domain.MetricCard, error)
	GetNodesMetric(ctx context.Context, duration string) (*domain.MetricCard, error)
	GetTenantsMetric(ctx context.Context, duration string) (*domain.MetricCard, error)
	GetCPUUtilization(ctx context.Context, duration string) (*domain.ResourceUtilization, error)
	GetMemoryUtilization(ctx context.Context, duration string) (*domain.ResourceUtilization, error)
	GetNodes(ctx context.Context) ([]domain.NodeMetric, error)
	GetNamespaces(ctx context.Context) ([]domain.NamespaceMetric, error)
	GetDatabaseIOMetrics(ctx context.Context, duration string) (*domain.DatabaseMetrics, error)
}
